README
Lingua-JA-FindDates version 0.01
================================

This module converts Japanese dates into English.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module doesn't require other modules or libraries.

COPYRIGHT AND LICENCE

Copyright (C) 2008 Ben Kasmin Bullock. All rights reserved.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public
License along with this program.  If not, see
<http://www.gnu.org/licenses/>.



Changes
0.028 2018-11-14

* Allow export of $jera and $jera2w
* Lower-case h/s etc. single-letter eras accepted (h 30 for Heisei 30, etc.)

0.027 2018-10-24

* Documentation

0.026 2017-10-17

* Documentation

0.025 2017-10-16

* Remove stray debugging code
* Documentation

0.024 2017-03-14

* Document the regjnums function
* Clean up of distribution
* Documentation errors etc.

0.023 2016-08-30

* Documentation errors corrected

0.022 2016-08-30

* Intervals between two era dates
* Change to machine-readable examples, better documentation

0.021 2016-08-29

* Add dates with a single kanji, like 明19.

0.020 2014-02-05

* Add README and MIN_PERL_VERSION in Makefile.PL.

0.019 2013-11-03

* Add squared symbols like ㍻ to the list of possibles.

0.018 2013-02-08

* Fix bug with Meiji start year (was 1870, now 1868).

0.017 2013-02-08

* Add two functions, "seireki_to_nengo" and "nengo_to_seireki".

0.016 2013-02-07

* Fix for tricky bug involving romaji.
* Code cleanup.

0.015 2012-07-27

* Move documentation to lib/Lingua/JA/FindDates.pod.

0.014 2012-07-27

* Add bugtracker, repository.

0.013 2012-03-19

* Add "subsjdate" script.

0.012 2012-03-19

* Documentation improvements.
* Change to ExtUtils::MakeMaker.
* Version control for source code.

0.011 2010-09-18

* Add meta files
* Handle apostrophe dates

0.010 2009-06-10

* Match only horizontal whitespace to prevent mismatches

0.009 2008-09-28

* Bug fix with "jun" disappearing sometimes.

0.008 2008-08-17

* Documentation fixes (see www.annocpan.org for details).
* Bug fixes (see rt.cpan.org for details).

0.007 2008-07-22

* Rewrote some of the documentation
* Added nyoro 0xFF5E ("Fullwidth tilde") for Microsoft Word nyoro.
* Changed to Module::Install.
* Stopped exporting kanji2number.

0.006 2008-07-17

* Fixed a bug with something like the second 日 in １月１日 日本人論
  being taken for Sunday.
* Added 元年 to regex
* Added note on bug 56902 in Perl.
* Changed interfaces for callbacks and for passing date information.
* Added intervals with weekdays like 平成９年１０月１７日（火）～２０日（金）
* Trimmed documentation.

0.005 2008-07-11

* Added support for ten day periods like 中旬 etc. 
* Added support for intervals like 平成１０年３月１～４日.

0.004 2008-07-08

* Fixed bugs with fiscal years like ２０００年度, turned off
  verboseness and added test to ensure that $verbose is switched off
  for release versions to avoid this happening again in the future.

0.003 2008-07-06

* Re-licensed to Perl licence. 

0.002 2008-07-04

* Fix mojibake in search.cpan.org by adding =encoding line at the top
  of the module.

0.001 2008-07-04

   - original version; created by h2xs 1.23 with options
       -X -n Lingua::JA::FindDates


testing/test.pl
#!/home/ben/software/install/bin/perl

use warnings;
use strict;
use utf8;
use lib '../lib/';
use Lingua::JA::FindDates 'subsjdate';
binmode STDOUT, "utf8";

my $test5 =<<EOF;
2008年07月03日 01:39:21 投稿
【西村修平・桜井誠編】平成20年7月2日毎日新聞社前抗議行動！
最低賃金法が大きく改正され、平成２０年７月１日から施行されました。
平成二十年七月四日
Opera 9.51(Windows版)を入れてみたけれども相變らず使ひ物にならない。何なんだらう。
日本義塾 平成二十年七月公開講義
公開日時：2008年06月28日 01時09分
更新日時：2008年06月29日 00時08分

◎【日本義塾七月公開講義】

　◎日　時　平成二十年七月二十五日（金）
　　　　　　午後六時半～九時（六時開場）
H20年7月壁紙
7月壁紙（1024×768）を作成しました（クリックすると拡大します）。
昭和４９年度
1999年度
EOF

#$Lingua::JA::FindDates::verbose = 1;

sub replace_callback
{
    my ($data, $jdate, $edate) = @_;
    print "$jdate -> $edate\n";
}
#print $test5;
subsjdate ($test5, {replace => \&replace_callback});


testing/test2.pl
#!/home/ben/software/install/bin/perl

use warnings;
use strict;
use utf8;
use lib '../lib/';
use Lingua::JA::FindDates 'subsjdate';
binmode STDOUT, "utf8";
my $has_newline =<<EOF;
平成二十年七月一日
    日本論・日本人論は非常に面白いものだ。
EOF

#$Lingua::JA::FindDates::verbose = 1;

print subsjdate($has_newline);


testing/test3.pl
#!/home/ben/software/install/bin/perl

use warnings;
use strict;
use utf8;
use lib './blib/lib/';
use Lingua::JA::FindDates 'subsjdate';
#$Lingua::JA::FindDates::verbose = 1;# 'subsjdate';
binmode STDOUT, "utf8";
# my @tests_interval = 
# ('昭和41年3月1〜12日',
#  '昭和41年3月1日〜12日',
#  '昭和41年1〜12月',
#  '昭和41年1月〜12月',
#  '昭和41年3月1日〜4月12日',
#  '昭和41年3月1日〜4月12日',
# '三月一日（木）〜３日（土）',
# '2008年7月一日（木）〜八月３日（土)',
# );

# for my $c (@tests_interval[0..1]) {
#     print "Looking for $c\n";
#     print $c, " ", subsjdate($c),"\n";
# }
# for my $c (@tests_interval[2..3]) {
#     print "Looking for $c\n";
#     print $c, " ", subsjdate($c),"\n";
# }
# for my $c (@tests_interval[4..5]) {
#     print "Looking for $c\n";
#     print $c, " ", subsjdate($c),"\n";
# }
# for my $c (@tests_interval[6..7]) {
#     print "Looking for $c\n";
#     print $c, " ", subsjdate($c),"\n";
# }
#'Thursday 1-Saturday 3 March'
#'Thursday July 1-Saturday August 3, 2008'
#print subsjdate ('2008年7月一日（木）〜３日（土)'),"\n";
print subsjdate ('平成９年１０月１７日（火）〜２０日（金）'),"\n";


make-pod.pl
#!/home/ben/software/install/bin/perl
use warnings;
use strict;
use Template;
use FindBin '$Bin';
use Perl::Build qw/get_commit get_info/;
use Perl::Build::Pod ':all';
use Deploy qw/do_system older/;
use Getopt::Long;
my $ok = GetOptions (
    'force' => \my $force,
    'verbose' => \my $verbose,
);
if (! $ok) {
    usage ();
    exit;
}
my %pbv = (base => $Bin, verbose => $verbose);
my $commit = get_commit (%pbv);
my $info = get_info (%pbv);
# Names of the input and output files containing the documentation.

my $pod = 'FindDates.pod';
my $input = "$Bin/lib/Lingua/JA/$pod.tmpl";
my $output = "$Bin/lib/Lingua/JA/$pod";

# Template toolkit variable holder

my %vars = (
    commit => $commit,
    info => $info,
);

my $tt = Template->new (
    ABSOLUTE => 1,
    INCLUDE_PATH => [
	$Bin,
	pbtmpl (),
	"$Bin/examples",
    ],
    ENCODING => 'UTF8',
    FILTERS => {
        xtidy => [
            \& xtidy,
            0,
        ],
    },
    STRICT => 1,
);

my @examples = <$Bin/examples/*.pl>;
for my $example (@examples) {
    my $output = $example;
    $output =~ s/\.pl$/-out.txt/;
    if (older ($output, $example) || $force) {
	do_system ("perl -I$Bin/blib/lib -I$Bin/blib/arch $example > $output 2>&1", $verbose);
    }
}

chmod 0644, $output;
$tt->process ($input, \%vars, $output, binmode => 'utf8')
    or die '' . $tt->error ();
chmod 0444, $output;

exit;

sub usage
{
print <<USAGEEOF;
--verbose
--force
USAGEEOF
}



lib/Lingua/JA/FindDates.pm
package Lingua::JA::FindDates;
use warnings;
use strict;
use Carp qw/carp croak cluck/;
use utf8;
use 5.010000;

require Exporter;
our @ISA = qw(Exporter);
our @EXPORT_OK= qw/subsjdate kanji2number seireki_to_nengo nengo_to_seireki
		   regjnums @jdatere $jera %jera2w/;
our %EXPORT_TAGS = (
    all => \@EXPORT_OK,
);

our $VERSION = '0.028';

# Kanji number conversion table.

my %kanjinums = 
(
    〇 => 0,
    一 => 1,
    二 => 2,
    三 => 3,
    四 => 4,
    五 => 5,
    六 => 6,
    七 => 7,
    八 => 8,
    九 => 9,
    十 => 10,
    百 => 100,
    # Dates shouldn't get any bigger than the following times a digit.
    千 => 1000,
); 

# The kanji digits.

my $kanjidigits = join ('', keys %kanjinums);

sub kanji2number
{
    my ($knum) = @_;
    return 1 if $knum eq '元';
    my @kanjis = split '', $knum;
    my $value = 0;
    my $keta = 1;
    while (1) {
	my $k = pop @kanjis;
	return $value if ! defined $k;
	my $val = $kanjinums{$k};
        # Make sure this kanji number is one we know how to handle.
	if (! defined $val) {
	    warn "can't cope with '$k' of input '$knum'";
	    return 0;
	}
        # If the value of the individual kanji is more than 10.
	if ($val >= 10) {
	    $keta = $val;
	    my $knext = pop @kanjis;
	    if (!$knext) {
		return $value + $val;
	    }
	    my $val_next = $kanjinums{$knext};
	    if (!defined $val_next) {
		warn "can't cope with '$knext' of input '$knum'.\n";
		return 0;
	    }
	    if ($val_next > 10) {
		push @kanjis, $knext;
		$value += $val;
	    }
            else {
		$value += $val_next * $val;
	    }
	}
        else {
            # $k is a kanji digit from 0 to 9, and $val is its value.
	    $value += $val * $keta;
	    $keta *= 10;
	}
    }
}

#  ____                               
# |  _ \ ___  __ _  _____  _____  ___ 
# | |_) / _ \/ _` |/ _ \ \/ / _ \/ __|
# |  _ <  __/ (_| |  __/>  <  __/\__ \
# |_| \_\___|\__, |\___/_/\_\___||___/
#            |___/                    

my $jdigit = qr/[０-９0-9]/;

# A regular expression to match Japanese numbers

my $jnumber = qr/($jdigit+|[$kanjidigits]+)/x;

# A regular expression to match a Western year

my $wyear = qr/
                  (
                      $jdigit{4}
                  |
                      [$kanjidigits]?千[$kanjidigits]*
                  |
                      [\']$jdigit{2}
                  )
                  \s*年
              /x;

my $alpha_era = qr/
                      # If the H, S, T, or M is part of a longer
                      # string of romaji, do not match it.
                      (?<![A-ZＡ-Ｚ])
                      (?:
                          [H|Ｈ|S|Ｓ|T|Ｔ|M|Ｍ]
                      )
                  /xi;

# The recent era names (Heisei, Showa, Taisho, Meiji). These eras are
# sometimes written using the letters H, S, T, and M.

our $jera = qr/($alpha_era|平|昭|大|明|平成|昭和|大正|明治|㍻|㍼|㍽|㍾)/;

# A map of Japanese eras to Western dates. These are the starting year
# of the period minus one, to allow for that the first year is "heisei
# one" rather than "heisei zero".

our %jera2w = (
    H    => 1988,
    Ｈ   => 1988,
    平成 => 1988,
    平 => 1988,
    '㍻' => 1988,
    S    => 1925,
    Ｓ   => 1925,
    昭和 => 1925,
    昭 => 1925,
    '㍼' => 1925,
    T    => 1911,
    Ｔ   => 1911,
    大正 => 1911,
    大 => 1911,
    '㍽' => 1911,
    M    => 1867,
    Ｍ   => 1867,
    明治 => 1867,
    明 => 1867,
    '㍾' => 1867,
);

# Japanese year, with era like "Heisei" at the beginning.

my $jyear = qr/
                  $jera
                  \h*
                  # Only match up to one or two of these digits, to
                  # prevent unlikely matches.
                  (
                      $jdigit{1,2}
                  |
                      [$kanjidigits]{1,2}
                  |
                      # The first year of an era, something like
                      # "昭和元年" (1926, the first year of the Showa era).
                      元 
                  )
                  \h*
                  年
              /x;

# The "jun" or approximately ten day periods (thirds of a month)

my %jun = qw/初 1 上 1 中 2 下 3/;

# The translations of the "jun" above into English.

my @jun2english = (
    'invalid',
    'early ',
    'mid-',
    'late ',
);

# Japanese days of the week, from Monday to Sunday.

my $weekdays = '月火水木金土日';
my @weekdays = split '',$weekdays;

# Match a string for a weekday, like 月曜日 or (日)
# The long part (?=\W) is to stop it from accidentally matching a
# kanji which is part of a different word, like the following:
#平成二十年七月一日
#    日本論・日本人論は非常に面白いものだ。

my $match_weekday =
    qr/[\（(]?
       ([$weekdays])
       (?:(?:(?:曜日|曜)[)\）])|[)\）]|(?=\W))
      /x;

# Match a day of the month, like 10日

my $match_dom = qr/$jnumber\h*日/;

# Match a month

my $match_month = qr/$jnumber\h*月/;

# Match a "jun" (a third of a month).

my $jun_keys = join ('', keys %jun);

my $match_jun = qr/([$jun_keys])\h*旬/;

# Match a month+jun

my $match_month_jun = qr/$match_month\h*$match_jun/;

# Match a month and day of month pair

my $match_month_day = qr/$match_month\h*$match_dom/;

# Match a Japanese year, month, day string

my $matchymd = qr/
                     $jyear
                     \h*
                     $match_month_day
                 /x;

# Match a Western year, month, day string

my $matchwymd = qr/$wyear\h*$match_month_day/;

# Match a Japanese year and month only

my $match_jyear_month = qr/$jyear\h*$match_month/;

# Match a Western year and month only

my $match_wyear_month = qr/$wyear\h*$match_month/;

# Match a month, day, weekday.

my $match_month_day_weekday = qr/$match_month_day\h*$match_weekday/;

# Separators used in date strings
# Microsoft Word uses Unicode 0xFF5E, the "fullwidth tilde", for nyoro symbol.

my $separators = qr/\h*[〜−~]\h*/;
 
#  _     _     _            __                                     
# | |   (_)___| |_    ___  / _|  _ __ ___  __ _  _____  _____  ___ 
# | |   | / __| __|  / _ \| |_  | '__/ _ \/ _` |/ _ \ \/ / _ \/ __|
# | |___| \__ \ |_  | (_) |  _| | | |  __/ (_| |  __/>  <  __/\__ \
# |_____|_|___/\__|  \___/|_|   |_|  \___|\__, |\___/_/\_\___||___/
#                                         |___/                    

# This a list of date regular expressions.

our @jdatere = (

    # Match an empty string like 平成 月 日 as found on a form etc.

    [qr/
           $jyear
           (\h+)
           月
           \h+
           日
       /x,
     "ejx"],

    # Add match for dummy strings here!

    # Match a Japanese era, year, 2 x (month day weekday) combination

    [qr/
           $matchymd
           \h*$match_weekday
           $separators
           $matchymd
           \h*$match_weekday
       /x,
     "e1j1m1d1w1e2j2m2d2w2"],

    # Match 2 x (era, year, month, day) combination

    [qr/
           $matchymd
           $separators
           $matchymd
       /x,
     "e1j1m1d1e2j2m2d2"],

    # Match a Japanese era, year, month 2 x (day, weekday) combination

    [qr/
           $matchymd
           $match_weekday
           $separators
           $match_dom
           \h*
           $match_weekday
       /x, 
     "ejmd1w1d2w2"],

    # Match a Japanese era, year, month 2 x day combination

    [qr/
           $matchymd
           $separators
           $match_dom
           \h*
           $match_weekday
       /x,
     "ejmd1d2"],

    # Match 2x(Western year, month, day, weekday) combination

    [qr/
           $matchwymd
           \h*
           $match_weekday
           $separators
           $matchwymd
           $match_weekday
       /x,
     "y1m1d1w1y2m2d2w2"],

    # Match a Western year, 2x(month, day, weekday) combination

    [qr/
           $matchwymd
           \h*
           $match_weekday
           $separators
           $match_month_day_weekday
       /x,
     "ym1d1w1m2d2w2"],

    # Match a Western year, month, 2x(day, weekday) combination

    [qr/
           $matchwymd
           \h*
           $match_weekday
           $separators
           $match_dom
           \h*
           $match_weekday
       /x,
     "ymd1w1d2w2"],

    # Match a Western year, month, 2x(day) combination

    [qr/
           $matchwymd
           $separators
           $match_dom
       /x,
     "ymd1d2"],

    # Match a Japanese era, year, month1 day1 - month 2 day2 combination

    [qr/
           $matchymd
           $separators
           $match_month_day
       /x,
     "ejm1d1m2d2"],

    # Match 2 x ( Japanese era, year, month) combination

    [qr/
           $jyear
           \h*
           $jnumber
           \h*月?
           $separators
           $jyear
           \h*
           $match_month
       /x, "e1j1m1e2j2m2"],

    # Match a Japanese era, year, month1 - month 2 combination

    [qr/
           $jyear
           \h*
           $jnumber
           \h*月?
           $separators
           $match_month
       /x, "ejm1m2"],

    # Match a Japanese era, year, month, day1 - day2 combination

    [qr/
           $match_jyear_month
           \h*
           $jnumber
           \h*日?
           $separators
           $match_dom
       /x,
     "ejmd1d2"],

    # Match a Japanese era, year, month, day, weekday combination

    [qr/
           $matchymd
           \h*
           $match_weekday
       /x,
     "ejmdw"],

    # Match a Japanese era, year, month, day

    [qr/$matchymd/,
     "ejmd"],

    # Match a Japanese era, year, month, jun

    [qr/
           $match_jyear_month
           \h*
           $match_jun
       /x,
     "ejmz"],

    # Match a Western year, month, day, weekday combination

    [qr/
           $matchwymd
           \h*
           $match_weekday
       /x,
     "ymdw"],

    # Match a Western year, month, day combination

    [qr/$matchwymd/,
     "ymd"],

    # Match a Western year, month, jun combination

    [qr/
           $match_wyear_month
           \h*
           $match_jun
       /x,
     "ymz"],

    # Match a Japanese era, year, month

    [qr/
           $jyear
           \h*
           $jnumber
           \h*
           月
       /x,
     "ejm"],

    # Match a Western year, month

    [qr/$match_wyear_month/,
     "ym"],

    # Match 2 x (month, day, weekday)

    [qr/
           $match_month_day_weekday
           $separators
           $match_month_day_weekday
       /x, 
     "m1d1w1m2d2w2"],

    # Match month, 2 x (day, weekday)

    [qr/
           $match_month_day_weekday
           $separators
           $match_dom
           \h*
           $match_weekday
       /x,
     "md1w1d2w2"],

    # Match month, 2 x (day, weekday)

    [qr/
           $match_month_day
           $separators
           $match_dom
       /x,
     "md1d2"],

    # Match a month, day, weekday

    [qr/$match_month_day_weekday/,
     "mdw"],

    # Match a month, day

    [qr/$match_month_day/,
     "md"],

    # Match a fiscal year (年度, nendo in Japanese). These usually don't
    # have months combined with them, so there is nothing to match a
    # fiscal year with a month.

    [qr/
           $jyear
           度
       /x,
     "en"],

    # Match a fiscal year (年度, nendo in Japanese). These usually don't
    # have months combined with them, so there is nothing to match a
    # fiscal year with a month.

    [qr/
           $wyear
           度
       /x,
     "n"],

    # Match a Japanese era, year

    [qr/$jyear/,
     "ej"],

    # Match a Western year

    [qr/$wyear/,
     "y"],

    # Match a month with a jun

    [
        qr/
              $match_month
              \h*
              $match_jun
          /x,
        "mz"
    ],

    # Match a month

    [
        qr/$match_month/,
        "m"
    ],
);

my @months = qw/Invalid
                January
                February
                March
                April
                May
                June
                July
		August
                September
                October
                November
                December
                MM/;

my @days = qw/Invalid
              Monday
              Tuesday
              Wednesday
              Thursday
              Friday
              Saturday
              Sunday/;

# This is a translation table from the Japanese weekday names to the
# English ones.

my %j2eweekday;

@j2eweekday{@weekdays} = (1..7);

# This is the default routine for turning a Japanese date into a
# foreign-style one.

sub make_date
{
    goto & default_make_date;
}

sub make_date_interval
{
    goto & default_make_date_interval;
}

sub default_make_date
{
    my ($datehash) = @_;
    my ($year, $month, $date, $wday, $jun) = 
	@{$datehash}{qw/year month date wday jun/};
    if (!$year && !$month && !$date && !$jun) {
	carp "No valid inputs\n";
	return;
    }
    my $edate = '';
    $edate = $days[$wday].", " if $wday;
    if ($month) {
	$month = int ($month); # In case it is 07 etc.
	$edate .= $months[$month];
	if ($jun) {
	    $edate = $jun2english[$jun] . $edate;
	}
    }
    if ($date) {
	$edate .= " " if length ($edate);
	$date = int ($date); # In case it is 07 etc.
	$date = "DD" if $date == 32;
	if ($year) {
	    $edate .= "$date, $year";
	}
	else {
	    $edate .= "$date";
	}
    }
    elsif ($year) {
	if (length ($edate) > 0) {
	    $edate .= " ";
	}
	$edate .= $year;
    }
    return $edate;
}

our $date_sep = '-';

# This is the default routine for turning a date interval into a
# foreign-style one, which is then substituted into the text.

sub default_make_date_interval
{
    my ($date1, $date2) = @_;
    my $einterval = '';
    my $usecomma;
    # The case of an interval with different years doesn't need to be
    # considered, because each date in that case can be considered a
    # single date.

    if ($date2->{month}) {
	if (!$date1->{month}) {
	    carp "end month but no starting month";
	    return;
	}
    }
    if ($date1->{month}) {
	if ($date1->{wday} && $date2->{wday}) {
	    if (! $date1->{date} || ! $date2->{date}) {
		carp "malformed date has weekdays but not days of month";
		return;
	    }
	    $usecomma = 1;
	    $einterval = $days[$date1->{wday}]  . " " . $date1->{date} .
		         ($date2->{month} ? ' '.$months[int ($date1->{month})] : ''). $date_sep .
		         $days[$date2->{wday}]  . " " . $date2->{date} . " " .
			 ($date2->{month} ? $months[int ($date2->{month})] : $months[int ($date1->{month})]);
	}
        elsif ($date1->{date} && $date2->{date}) {
	    $usecomma = 1;
	    if ($date1->{wday} || $date2->{wday}) {
		carp "malformed date interval: ",
		    "has weekday for one date $date1->{wday} but not the other one $date2->{wday} .";
		return;
	    }
	    $einterval = $months[int ($date1->{month})] . ' ' .
		         $date1->{date} . $date_sep .
			 ($date2->{month} ? 
			  $months[int ($date2->{month})] . ' ' : '') .
		         $date2->{date};
	}
        else { # no dates or weekdays
	    if ($date1->{date} || $date2->{date}) {
		cluck "malformed date interval: only one day of month";
		return;
	    }
	    if (!$date2->{month}) {
		carp "start month but no end month or date";
		return;
	    }
	    $einterval = $months[int($date1->{month})] . $date_sep . 
		         $months[int($date2->{month})] .
			 $einterval;
	}
    }
    else { # weekday - day / weekday - day case.
	if ($date1->{wday} && $date2->{wday}) {
	    if (! $date1->{date} || ! $date2->{date}) {
		carp "malformed date has weekdays but not days of month";
		return;
	    }
	    $einterval = $date1->{wday}  . " " . $date1->{date} . $date_sep .
		         $date2->{wday}  . " " . $date2->{date};
	}
    }
    if ($date1->{year}) {
	my $year1 = ($usecomma ? ', ': ' ').$date1->{year};
	if (! $date2->{year} || $date2->{year} == $date1->{year}) {
	    $einterval .= $year1;
	}
	else {
	    $einterval =~ s/\Q$date_sep/$year1$date_sep/;
	    my $year2 = ($usecomma ? ', ': ' ').$date2->{year};
	    $einterval .= $year2;
	}
    }
    return $einterval;
}

our $verbose = 0;

sub subsjdate
{
    # $text is the text to substitute. It needs to be in Perl's
    # internal encoding.
    # $replace_callback is a routine to call back if we find valid dates.
    # $data is arbitrary data to pass to the callback routine.
    my ($text, $c) = @_;
    # Save doing existence tests.
    if (! $c) {
        $c = {};
    }
    if (! $text) {
        return $text;
    }
    # Loop through all the possible regular expressions.
    for my $datere (@jdatere) {
	my $regex = $datere->[0];
	my @process = split (/(?=[a-z][12]?)/, $datere->[1]);
        if ($verbose) {
            print "Looking for $datere->[1]\n";
        }
	while ($text =~ /($regex)/g) {
	    my $date1;
	    my $date2;
            # The matching string is in the following variable.
	    my $orig = $1;
	    my @matches = ($2,$3,$4,$5,$6,$7,$8,$9);
            if ($verbose) {
                print "Found '$orig': ";
            }
	    for (0..$#matches) {
		my $arg = $matches[$_];

		last if !$arg;
		$arg =~ tr/０-９/0-9/;
		$arg =~ s/([$kanjidigits]+|元)/kanji2number($1)/ge;
                if ($verbose) {
                    print "Arg $_: $arg ";
                }
		my $argdo = $process[$_];
		if ($argdo eq 'e1') { # Era name in Japanese
		    $date1->{year} = $jera2w{$arg};
		}
                elsif ($argdo eq 'j1') { # Japanese year
		    $date1->{year} += $arg;
		}
                elsif ($argdo eq 'y1') {
		    $date1->{year} = $arg;
		}
		elsif ($argdo eq 'e2') { # Era name in Japanese
		    $date2->{year} = $jera2w{$arg};
		}
                elsif ($argdo eq 'j2') { # Japanese year
		    $date2->{year} += $arg;
		}
                elsif ($argdo eq 'y2') {
		    $date2->{year} = $arg;
		}
		elsif ($argdo eq 'e') { # Era name in Japanese
		    $date1->{year} = $jera2w{$arg};
		}
                elsif ($argdo eq 'j') { # Japanese year
		    $date1->{year} += $arg;
		}
                elsif ($argdo eq 'y') {
		    $date1->{year} = $arg;
		}
                elsif ($argdo eq 'n') {
		    $date1->{year} += $arg;
		    $date1->{year} = "fiscal ".$date1->{year};
		}
                elsif ($argdo eq 'm' || $argdo eq 'm1') {
		    $date1->{month} = $arg;
		}
                elsif ($argdo eq 'd' || $argdo eq 'd1') {
		    $date1->{date} = $arg;
		}
                elsif ($argdo eq 'm2') {
		    $date2->{month} = $arg;
		}
                elsif ($argdo eq 'd2') {
		    $date2->{date} = $arg;
		}
                elsif ($argdo eq 'w' || $argdo eq 'w1') {
		    $date1->{wday} = $j2eweekday{$arg};
		}
                elsif ($argdo eq 'w2') {
		    $date2->{wday} = $j2eweekday{$arg};
		}
                elsif ($argdo eq 'z') {
		    $date1->{jun} = $jun{$arg};
		}
                elsif ($argdo eq 'x') {
                    if ($verbose) {
                        print "Dummy date '$orig'.\n";
                    }
		    $date1->{date}  = 32;
		    $date1->{month} = 13;
		}
	    }
	    my $edate;
	    if ($date2) {
                # Date interval
		if ($c->{make_date_interval}) {
		    $edate = &{$c->{make_date_interval}} ($c->{data}, $orig,
							  $date1, $date2);
		}
                else {
		    $edate = default_make_date_interval ($date1, $date2);
		}
	    }
            else {
                # Single date
		if ($c->{make_date}) {
		    $edate = &{$c->{make_date}}($c->{data}, $orig, $date1);
		}
                else {
		    $edate = default_make_date ($date1);
		}
	    }
            if ($verbose) {
                print "-> '$edate'\n";
            }
	    $text =~ s/\Q$orig\E/$edate/g;
	    if ($c->{replace}) {
		&{$c->{replace}} ($c->{data}, $orig, $edate);
	    }
	}
    }
    return $text;
}

sub nengo_to_seireki
{
    my ($text) = @_;
    my %data;
    $data{count} = 0;

    my $out_text = subsjdate (
        $text, {
            make_date => \& nengo_to_seireki_make_date, 
            data => \%data,
        }
    );
    $out_text =~ s/#REPLACEME(\d+)REPLACEME#/$data{$1}/g;
    return $out_text;
}

sub nengo_to_seireki_make_date
{
    my ($data, $original, $date) = @_;
    if ($date->{year}) {
        $original =~ s/.*年/$date->{year}年/;
        my $count = $data->{count};
        $data->{$count} = $original;
        $data->{count}++;
        return "#REPLACEME${count}REPLACEME#";
    }
    else {
        return $original;
    }
}

sub seireki_to_nengo
{
    my ($text) = @_;
    my %data;
    $data{count} = 0;

    my $out_text = subsjdate (
        $text, {
            make_date => \& seireki_to_nengo_make_date, 
            data => \%data,
        }
    );
    $out_text =~ s/#REPLACEME(\d+)REPLACEME#/$data{$1}/g;
    return $out_text;
}

sub seireki_to_nengo_make_date
{
    my ($data, $original, $date) = @_;
    my $year = $date->{year};
    my @eras = (
        ['平成', 1989, 1, 8],
        ['昭和', 1926, 12, 25],
        ['大正', 1912, 7, 30],
        ['明治', 1868, 1, 25],
    );
    if (defined $year) {
        for my $era (@eras) {
            my $ename = $era->[0];
            my $eyear = $era->[1];
            my $emonth = $era->[2];
            my $eday = $era->[3];
            my $month = $date->{month};
            my $date = $date->{date};

            # This is a flag which says whether to perform a
            # substitution of the year or not.

            my $subs;

            # If the year is greater than the era year, or if the year
            # is the same as the era year and we do not know the
            # month, just replace.

            if ($year > $eyear ||
                ($year == $eyear && ! defined ($month))) {
                $subs = 1;
            }

            # If the year is the same, and there is a month

            elsif ($year == $eyear && defined ($month)) {

                # If there is a day of the month, then only substitute
                # if the month is greater than the changeover month,
                # or the month is the same, and the day of the month
                # is greater than or equal to the changeover day of
                # the month.

                if (defined ($date)) {
                    if ($month > $emonth ||
                        ($month == $emonth && $date >= $eday)) {
                        $subs = 1;
                    }
                }

                # If we don't know the day of the month, substitute if
                # the month is greater than or equal to the changeover
                # month.

                elsif ($month >= $emonth) {
                    $subs = 1;
                }
            }
            if ($subs) {

                # Only substitute if we need to.

                if ($original !~ /$ename/) {

                    # The year counting starts from 1, so we add 1 to
                    # the difference.

                    my $hyear = $year - $eyear + 1;
                    $original =~ s/\d+年/$ename${hyear}年/;
                }

                # Don't replace again, stop the loop.

                last;
            }
        }
    }
    my $count = $data->{count};
    $data->{$count} = $original;
    $data->{count}++;

    # This is a tag for substituting with.

    return "#REPLACEME${count}REPLACEME#";
}

# Regularize any small integer Japanese numbers in a piece of text.

sub regjnums
{
    my ($input) = @_;
    $input =~ tr/０-９/0-9/;
    $input =~ s/([$kanjidigits]+)/kanji2number($1)/ge;
    return $input;
}

1;

